File: tests/language-support/lua/storage.lua
local Storage = {}

function Storage.add_book(shelf, book)
    if Storage.is_full(shelf) then
        error("Shelf '" .. shelf.name .. "' is at capacity (" .. shelf.capacity .. ")")
    end
    table.insert(shelf.books, book)
end

function Storage.remove_book(shelf, isbn)
    for i, book in ipairs(shelf.books) do
        if book.isbn == isbn then
            return table.remove(shelf.books, i)
        end
    end
    error("No book with ISBN '" .. isbn .. "' found on shelf")
end

function Storage.find_by_author(shelf, author)
    local query = string.lower(author)
    local results = {}
    for _, book in ipairs(shelf.books) do
        if string.find(string.lower(book.author), query, 1, true) then
            table.insert(results, book)
        end
    end
    return results
end

function Storage.find_by_year_range(shelf, start_year, end_year)
    local results = {}
    for _, book in ipairs(shelf.books) do
        if book.year >= start_year and book.year <= end_year then
            table.insert(results, book)
        end
    end
    return results
end

function Storage.sort_by_title(shelf)
    local sorted = {}
    for _, book in ipairs(shelf.books) do
        table.insert(sorted, book)
    end
    table.sort(sorted, function(a, b) return a.title < b.title end)
    return sorted
end

function Storage.is_full(shelf)
    return #shelf.books >= shelf.capacity
end

function Storage.generate_report(shelf)
    local books = shelf.books
    local total = #books
    local available_count = 0
    for _, book in ipairs(books) do
        if book.available then available_count = available_count + 1 end
    end
    local avail_pct = total > 0 and math.floor(available_count * 100 / total) or 0
    local cap_pct = shelf.capacity > 0 and math.floor(total * 100 / shelf.capacity) or 0

    local author_counts = {}
    local author_order = {}
    for _, book in ipairs(books) do
        if not author_counts[book.author] then
            table.insert(author_order, book.author)
            author_counts[book.author] = 0
        end
        author_counts[book.author] = author_counts[book.author] + 1
    end
    local unique_authors = #author_order

    local min_year, max_year = 0, 0
    if total > 0 then
        min_year = books[1].year
        max_year = books[1].year
        for _, book in ipairs(books) do
            if book.year < min_year then min_year = book.year end
            if book.year > max_year then max_year = book.year end
        end
    end

    local lines = {}
    table.insert(lines, "=== Library Report: " .. shelf.name .. " ===")
    table.insert(lines, "Total books: " .. total)
    table.insert(lines, "Available: " .. available_count .. " / " .. total .. " (" .. avail_pct .. "%)")
    table.insert(lines, "Capacity: " .. total .. " / " .. shelf.capacity .. " (" .. cap_pct .. "% full)")
    table.insert(lines, "")

    table.insert(lines, "Authors (" .. unique_authors .. " unique):")
    for _, author in ipairs(author_order) do
        table.insert(lines, "  - " .. author .. " (" .. author_counts[author] .. " books)")
    end
    table.insert(lines, "")

    table.insert(lines, "Year range: " .. min_year .. " - " .. max_year)
    table.insert(lines, "")

    table.insert(lines, "Books by availability:")
    for _, book in ipairs(books) do
        local marker = book.available and "+" or "-"
        table.insert(lines, "  [" .. marker .. "] " .. book.title .. " by " .. book.author .. " (" .. book.year .. ") - " .. book.isbn)
    end

    return table.concat(lines, "\n")
end

return Storage

File: tests/language-support/erlang/src/storage.erl
%% @doc Storage operations for managing books on shelves.
-module(storage).
-export([add_book/2, remove_book/2, find_by_author/2, find_by_year_range/3,
         sort_by_title/1, is_full/1, generate_report/1]).

%% @doc Check whether the shelf is at capacity.
-spec is_full(models:shelf()) -> boolean().
is_full(#{capacity := Capacity, books := Books}) ->
    length(Books) >= Capacity.

%% @doc Add a book to the shelf.
-spec add_book(models:shelf(), models:book()) -> {ok, models:shelf()} | {error, string()}.
add_book(Shelf = #{name := Name, capacity := Capacity, books := Books}, Book) ->
    case is_full(Shelf) of
        true ->
            {error, lists:flatten(io_lib:format("Shelf '~s' is at capacity (~B)", [Name, Capacity]))};
        false ->
            {ok, Shelf#{books => Books ++ [Book]}}
    end.

%% @doc Remove a book by ISBN.
-spec remove_book(models:shelf(), string()) -> {ok, models:shelf(), models:book()} | {error, string()}.
remove_book(Shelf = #{books := Books}, Isbn) ->
    case lists:splitwith(fun(B) -> maps:get(isbn, B) =/= Isbn end, Books) of
        {_Before, []} ->
            {error, lists:flatten(io_lib:format("No book with ISBN '~s' found on shelf", [Isbn]))};
        {Before, [Found | After]} ->
            {ok, Shelf#{books => Before ++ After}, Found}
    end.

%% @doc Find books by author (case-insensitive substring match).
-spec find_by_author(models:shelf(), string()) -> [models:book()].
find_by_author(#{books := Books}, Query) ->
    LowerQuery = string:lowercase(Query),
    lists:filter(fun(B) ->
        LowerAuthor = string:lowercase(maps:get(author, B)),
        string:find(LowerAuthor, LowerQuery) =/= nomatch
    end, Books).

%% @doc Find books within an inclusive year range.
-spec find_by_year_range(models:shelf(), integer(), integer()) -> [models:book()].
find_by_year_range(#{books := Books}, StartYear, EndYear) ->
    lists:filter(fun(B) ->
        Y = maps:get(year, B),
        Y >= StartYear andalso Y =< EndYear
    end, Books).

%% @doc Return books sorted by title without mutating.
-spec sort_by_title(models:shelf()) -> [models:book()].
sort_by_title(#{books := Books}) ->
    lists:sort(fun(A, B) -> maps:get(title, A) =< maps:get(title, B) end, Books).

%% @doc Generate a formatted report for the shelf.
-spec generate_report(models:shelf()) -> string().
generate_report(#{name := Name, capacity := Capacity, books := Books}) ->
    Total = length(Books),
    AvailCount = length(lists:filter(fun(B) -> maps:get(available, B) end, Books)),
    AvailPct = case Total > 0 of true -> (AvailCount * 100) div Total; false -> 0 end,
    CapPct = case Capacity > 0 of true -> (Total * 100) div Capacity; false -> 0 end,
    AuthorCounts = count_authors(Books),
    UniqueCount = length(AuthorCounts),
    Years = [maps:get(year, B) || B <- Books],
    MinYear = lists:min(Years),
    MaxYear = lists:max(Years),
    Header = io_lib:format("=== Library Report: ~s ===", [Name]),
    AuthorLines = lists:map(fun({A, C}) ->
        io_lib:format("  - ~s (~B books)", [A, C])
    end, AuthorCounts),
    BookLines = lists:map(fun(B) ->
        Marker = case maps:get(available, B) of true -> "+"; false -> "-" end,
        io_lib:format("  [~s] ~s by ~s (~B) - ~s",
            [Marker, maps:get(title, B), maps:get(author, B),
             maps:get(year, B), maps:get(isbn, B)])
    end, Books),
    Lines = [Header,
             io_lib:format("Total books: ~B", [Total]),
             io_lib:format("Available: ~B / ~B (~B%)", [AvailCount, Total, AvailPct]),
             io_lib:format("Capacity: ~B / ~B (~B% full)", [Total, Capacity, CapPct]),
             "",
             io_lib:format("Authors (~B unique):", [UniqueCount])]
            ++ AuthorLines ++
            ["",
             io_lib:format("Year range: ~B - ~B", [MinYear, MaxYear]),
             "",
             "Books by availability:"]
            ++ BookLines,
    lists:flatten(lists:join("\n", Lines)).

%% @doc Count authors preserving insertion order.
count_authors(Books) ->
    count_authors(Books, Books, [], []).

count_authors([], _All, _Seen, Acc) ->
    lists:reverse(Acc);
count_authors([B | Rest], All, Seen, Acc) ->
    Author = maps:get(author, B),
    case lists:member(Author, Seen) of
        true -> count_authors(Rest, All, Seen, Acc);
        false ->
            AuthorCount = length(lists:filter(fun(X) -> maps:get(author, X) =:= Author end, All)),
            count_authors(Rest, All, [Author | Seen], [{Author, AuthorCount} | Acc])
    end.
